src/infra/flow-system/postgres/src/flow_system_event_store_postgres.rs

```diff
@@ -0,0 +1,257 @@
+                    r#"
+INSERT INTO dataset_flow_configuration_events (dataset_id, dataset_flow_type, event_type, event_time, event_payload)
+"#,
+                );
+
+                query_builder.push_values(events, |mut b, event| {
+                    b.push_bind(fk_dataset.dataset_id.to_string());
+                    b.push_bind(fk_dataset.flow_type);
+                    b.push_bind(event.typename());
+                    b.push_bind(event.event_time());
+                    b.push_bind(serde_json::to_value(event).unwrap());
+                });
+
+                query_builder
+            }
+            FlowKey::System(fk_system) => {
+                let mut query_builder = QueryBuilder::<Postgres>::new(
+                    r#"
+INSERT INTO system_flow_configuration_events (system_flow_type, event_type, event_time, event_payload)
+"#,
+                );
+
+                query_builder.push_values(events, |mut b, event| {
+                    b.push_bind(fk_system.flow_type);
+                    b.push_bind(event.typename());
+                    b.push_bind(event.event_time());
+                    b.push_bind(serde_json::to_value(event).unwrap());
+                });
+
+                query_builder
+            }
+        };
+
+        query_builder.push("RETURNING event_id");
+
+        #[derive(FromRow)]
+        struct ResultRow {
+            event_id: i64,
+        }
+
+        let rows = query_builder
+            .build_query_as::<ResultRow>()
+            .fetch_all(connection_mut)
+            .await
+            .int_err()?;
+
+        if let Some(last_row) = rows.last() {
+            Ok(EventID::new(last_row.event_id))
+        } else {
+            let event_id = i64::try_from(self.len().await.int_err()?).int_err()?;
+
+            Ok(EventID::new(event_id))
+        }
+    }
+
+    async fn len(&self) -> Result<usize, InternalError> {
+        let mut tr = self.transaction.lock().await;
+        let connection_mut = tr.connection_mut().await?;
+
+        let result = sqlx::query!(
+            r#"
+SELECT
+    CASE
+        WHEN is_called THEN last_value
+        ELSE 0
+    END AS count
+FROM flow_configuration_event_id_seq
+"#,
+        )
+        .fetch_one(connection_mut)
+        .await
+        .int_err()?;
+
+        let count = usize::try_from(result.count.unwrap()).int_err()?;
+
+        Ok(count)
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl FlowConfigurationEventStore for FlowSystemEventStorePostgres {
+    async fn list_all_dataset_ids(&self) -> FailableDatasetIDStream<'_> {
+        let mut tr = self.transaction.lock().await;
+
+        Box::pin(async_stream::stream! {
+            let connection_mut = tr.connection_mut().await?;
+
+            let mut query_stream = sqlx::query!(
+                r#"
+SELECT DISTINCT dataset_id
+FROM dataset_flow_configuration_events
```

Comment:
I think you can safely filter a certain type of event, like "created" to optimize